### PR TITLE
feat: stan supports token auth

### PR DESCRIPTION
**Pull request by whynowy**
Signed-off-by: Derek Wang <whynowy@gmail.com>

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -81,6 +81,7 @@ jobs:
           path: /home/runner/go/pkg/mod
           key: GOMODCACHE-v2-${{ hashFiles('**/go.mod') }}
       - run: make controller
+      - run: make testapi
```
Comment (alexec) [inline .github/workflows/build.yml:84]: minor - not needed - this can be pulled now

File: api/v1alpha1/stan.go [hunk 2]
```diff
@@ -8,4 +19,18 @@ type STAN struct {
 	ClusterID         string        `json:"clusterId,omitempty" protobuf:"bytes,5,opt,name=clusterId"`
 	Subject           string        `json:"subject" protobuf:"bytes,3,opt,name=subject"`
 	SubjectPrefix     SubjectPrefix `json:"subjectPrefix,omitempty" protobuf:"bytes,6,opt,name=subjectPrefix,casttype=SubjectPrefix"`
+	Auth              *STANAuth     `json:"auth,omitempty" protobuf:"bytes,7,opt,name=auth"`
```
Comment (alexec) [inline api/v1alpha1/stan.go:22]: *  does not need to be a pointer
* use kubebuilder:default to set the default

Note that even if you set a default, you should cater for empty string

File: config/ci/controller-manager-deploy.yaml [hunk 3]
```diff
@@ -10,5 +10,5 @@ spec:
       containers:
         - name: manager
           env:
-            - name: PULL_POLICY
+            - name: ARGO_DATAFLOW_PULL_POLICY
```
Comment (alexec) [inline config/ci/controller-manager-deploy.yaml:13]: ops!

File: runner/sidecar/sinks.go [hunk 4]
```diff
@@ -108,8 +109,30 @@ func connectKafkaSink(x *dfv1.Kafka, sinkName string) (func(msg []byte) error, e
 }
 
 func connectSTANSink(sinkName string, x *dfv1.STAN) (func(msg []byte) error, error) {
+	opts := []nats.Option{}
+	switch x.AuthStrategy() {
+	case dfv1.STANAuthToken:
+		token, err := getSTANAuthToken(context.Background(), x)
+		if err != nil {
+			return nil, err
```
Comment (alexec) [inline runner/sidecar/sinks.go:117]: minor -  I like to always wrap my errors so I can add extra diagnostics, e.g.

```
fmt.Errorf("failed to get STAN token %q: %w", sinkName, err)
```

File: runner/sidecar/sinks.go [hunk 5]
```diff
@@ -108,8 +109,30 @@ func connectKafkaSink(x *dfv1.Kafka, sinkName string) (func(msg []byte) error, e
...
+		}
+		opts = append(opts, nats.Token(token))
+	default:
+	}
+	logger.Info("nats auth strategy: "+string(x.AuthStrategy()), "sink", sinkName)
```
Comment (alexec) [inline runner/sidecar/sinks.go:122]: minor - use structured logging

File: runner/sidecar/sinks.go [hunk 6]
```diff
@@ -108,8 +109,30 @@ func connectKafkaSink(x *dfv1.Kafka, sinkName string) (func(msg []byte) error, e
...
+	nc, err := nats.Connect(x.NATSURL, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to nats url=%s subject=%s: %w", x.NATSURL, x.Subject, err)
```
Comment (alexec) [inline runner/sidecar/sinks.go:125]: minor - use %q rather than %s so that it is quoted

File: runner/sidecar/sources.go [hunk 7]
```diff
@@ -196,8 +197,30 @@ func registerKafkaSetPendingHook(x *dfv1.Kafka, sourceName string, client sarama
...
 func connectSTANSource(ctx context.Context, sourceName string, x *dfv1.STAN, f func(ctx context.Context, msg []byte) error) error {
+	opts := []nats.Option{}
+	switch x.AuthStrategy() {
+	case dfv1.STANAuthToken:
+		token, err := getSTANAuthToken(ctx, x)
+		if err != nil {
+			return err
+		}
+		opts = append(opts, nats.Token(token))
+	default:
```
Comment (alexec) [inline runner/sidecar/sources.go:208]: minor - empty default can be removed

File: runner/sidecar/sources.go [hunk 8]
```diff
@@ -196,8 +197,30 @@ func registerKafkaSetPendingHook(x *dfv1.Kafka, sourceName string, client sarama
...
+	}
+	logger.Info("nats auth strategy: "+string(x.AuthStrategy()), "source", sourceName)
+	nc, err := nats.Connect(x.NATSURL, opts...)
+	if err != nil {
+		return fmt.Errorf("failed to connect to nats url=%s subject=%s: %w", x.NATSURL, x.Subject, err)
+	}
+	beforeClosers = append(beforeClosers, func(ctx context.Context) error {
```
Comment (alexec) [inline runner/sidecar/sources.go:215]: I think I'll refactor this and add `addShutdownHook(hook, priority)`

**Review by alexec — APPROVED** [review]

File: .github/workflows/build.yml (hunk 1 above)
Comment (whynowy) [reply .github/workflows/build.yml:84]: It's needed when `testapi` has code changes.

File: api/v1alpha1/stan.go (hunk 2 above)
Comment (whynowy) [reply api/v1alpha1/stan.go:22]: I always feel weird when I do not set any value for a filed in a spec, but it gets a value after applying it to k8s.

**Pull request merged**